### ISSUE TITLE
(let) requires explicit type annotations for struct accessors

### ISSUE THREAD

**Issue by wargrey** [racket/typed-racket#415]
### What version of Racket are you using?

I am not sure what exactly the broken condition is. It also happens to occur in version 6.5 in few (normal) situations. In version 6.6, it occurs more often, almost every newly written `let`s require explicit type annotations for struct accessors.
### What program did you run?

```
(struct selector ([ids : (Listof Keyword)]) #:prefab)
(struct subject ([id : (U Keyword (Pairof Keyword (Listof Keyword)))]) #:prefab)

(define s (selector '(#:id)))
(define e (subject '#:123))

;;; Okay
(let ([ids : (Listof Keyword) (selector-ids s)]
      [id : (U Keyword (Pairof Keyword (Listof Keyword))) (subject-id e)])
  (cond [(null? ids) #true]
        [(keyword? id) (and (null? (cdr ids)) (eq? (car ids) id))]
        [else (set=? (list->set ids) (list->set id))]))

;;; Broken
(let ([ids (selector-ids s)]
      [id (subject-id e)])
  (cond [(null? ids) #true]
        [(keyword? id) (and (null? (cdr ids)) (eq? (car ids) id))]
        [else (set=? (list->set ids) (list->set id))]))

;;; Okay
(define ids (selector-ids s))
(define id (subject-id e))
(cond [(null? ids) #true]
      [(keyword? id) (and (null? (cdr ids)) (eq? (car ids) id))]
      [else (set=? (list->set ids) (list->set id))])
```
### What should have happened?

All three test cases should be evaluated to `false`. 
### If you got an error message, please include it here.

No more error messages.

**Comment by wargrey** [comment]
Same behavior shows up within `in-list`.
But it seems that old code has not been found broken.

**Comment by andmkent** [comment]
Minor update -- the behavior is correct when the structs are not prefab.

**Comment by wargrey** [comment]
Oh, that is informative.
All structs in my current application are prefab structs because I need the serialization mechanism.

**Comment by wargrey** [comment]
@pnwamk

Hello, what's the status of this issue?
This is really an awkward thing, since prefab struct is the most efficient serialization mechanism that out of box in Typed Racket, it can play an important role in in-process communication.

Now the big problem is, `match` loses all the type information about prefab structures, even worse, `assert` deals nothing with it without following a silly `(define v : Any instance)` or `(define v : Any (struct-field instance))`

Thanks in advance.

**Comment by wargrey** [comment]
```racket
#lang typed/racket

(module message typed/racket
  (provide (all-defined-out))
  
  (struct message ([level : Log-Level] [brief : String] [detail : Any] [topic : Symbol]) #:prefab))

(module test typed/racket
  (provide (all-defined-out))

  (require (submod ".." message))
  
  (define main : (-> Place Void)
    (lambda [chin]
      (match (place-channel-get chin)
        [(cons (? symbol? topic) (and (message level brief detail topic) msg))
         ;(define silly-but-okay : Any brief)
         (displayln (string-append (assert brief string?) ": " (~a (message-detail msg))))
         (void '(assert silly-but-okay string?))]
        [_ (void 'error)]))))
```

here is an example, it seems that if the `match` is used in the defining module, there is no such issue.

**Comment by pnwamk** [comment]
I should have some time today (or late late tonight) to dig into this and I'll get back to you (hopefully with a fix!).

**Comment by pnwamk** [comment]
I believe this fixes the issues: https://github.com/racket/typed-racket/pull/500

**Comment by pnwamk** [comment]
Fixed by 7599aedf2c8ea354d1b4965ec129ea4dd85f8010.
